Fix Netatmo camera name does not show under Media -> Media sources -> Camera

codyc1515 opened the pull request:
## Proposed change
Fixes issue #105268 where Netatmo camera name does not show under Media -> Media sources -> Camera

### Before
![image](https://github.com/home-assistant/core/assets/50791984/c83abbfc-d479-4329-ac48-1950ce31a714)

### After
![image](https://github.com/home-assistant/core/assets/50791984/85f4dba6-8541-44bb-bea8-b38a209f335f)

## Type of change
- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: fixes #105268 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

codyc1515 committed:
Fixes issue where Netatmo camera name does not show under Media -> Media sources ->Camera

Fixes #105268

codyc1515 commented:
Not sure why the CI is failing (appears to be unrelated).

cgtobi commented:
It is directly related since the change directly influences the entity names.

cgtobi commented:
This change also has effects on the camera naming throughout the app. Did you check that? And yes, if we want to keep this change then the tests need to be updated. I currently wonder if this is the right fix for this issue since you described in the issue that other cameras also have naming related issues. Did I understand that correctly?

codyc1515 commented:
Some other integrations also had issues (IMO), such as taking the name from the config entry or the default name of the camera entity (even if the user customised it already).

That is different to what we are discussing here (and new issues can be raised for those) because in the case of Netatmo, there is simply no name showing. So, it’s not a question of breaking changes nor arguing a preference of where you’d prefer the name to come from.

cgtobi commented:
Well, as I said, this might fix the missing names under media/cameras but would also alter the camera names in other places.

codyc1515 commented:
I didn’t experience that. The name appears to be consistent everywhere. This just brings everything in line.

cgtobi commented:
How many cameras do you have in your system? I only can tell what I see here in my dev environment where I have multiple cameras and the tests prove it as well. So in my opinion this fix is not complete yet.

joostlek commented on homeassistant/components/netatmo/camera.py:
```diff
@@ -97,6 +97,7 @@ def __init__(
         self._camera = cast(NaModules.Camera, netatmo_device.device)
         self._id = self._camera.entity_id
         self._home_id = self._camera.home.entity_id
+        self._attr_name = self._camera.name
```
Does the camera have a device attached? If the device name is the camera name, please set `_attr_name = None` as it will then use the device name as entity name

codyc1515 replied:
The device name _is_ the camera name. On checking, this explains why the name was duplicating. Thanks for the tip - I am testing it now.

codyc1515 replied:
Unfortunately, that did not work - back to showing a null value in the Camera screen. This is what I put:

```
self._camera = cast(NaModules.Camera, netatmo_device.device)
self._id = self._camera.entity_id
self._home_id = self._camera.home.entity_id
self._attr_name = None
self._device_name = self._camera.name
self._model = self._camera.device_type
self._config_url = CONF_URL_SECURITY
self._attr_unique_id = f"{self._id}-{self._model}"
self._quality = DEFAULT_QUALITY
self._monitoring: bool | None = None
self._light_state = None
```

cgtobi replied:
_attr_name is set to None in the Netatmo base class.

joostlek replied:
I mean, if I understand correctly, the name duplicating is wrong.

But isn't this just a bug in the media browser then instead of a bug in the entity name?

cgtobi replied:
Exactly my understanding.

codyc1515 commented:
> How many cameras do you have in your system? I only can tell what I see here in my dev environment where I have multiple cameras and the tests prove it as well. So in my opinion this fix is not complete yet.

Used to have two (one each of indoor and outdoor) but the outdoor one infrared died, so I returned it and got a Eufy (which is much worse to use in HA and HK, unfortunately).

codyc1515 commented:
Have we reached a consensus on the way to tackle this?

cgtobi commented:
I still don't agree that this is the solution and the tests are failing.

cgtobi commented:
From what you found out it looks to me that the camera media source is not respecting the device names properly but I have not yet been able to track down where this needs to be fixed.

cgtobi commented on homeassistant/components/netatmo/camera.py:
```diff
@@ -97,6 +97,7 @@ def __init__(
         self._camera = cast(NaModules.Camera, netatmo_device.device)
         self._id = self._camera.entity_id
         self._home_id = self._camera.home.entity_id
+        self._attr_name = self._camera.name
         self._device_name = self._camera.name
```
```suggestion
        self._device_name = self._attr_name = self._camera.name
```

And please remove line 86 `_attr_has_entity_name = True`

codyc1515 committed:
Remove entity name and change has_entity_name to False

has_entity_name has to be retained (per https://developers.home-assistant.io/docs/core/entity/#has_entity_name-true-mandatory-for-new-integrations)

cgtobi reviewed (approved):
Thanks @codyc1515 LGTM

codyc1515 commented:
I have implemented all the requested changes - I am very grateful for all your feedback. This is the best yet, with all tests passing. There is no duplication of "Device Name + Entity Name" anymore and it's showing fine in the regular UI vs. Camera media sources UI.

That said, is it expected that if the entity name for the camera is customised that it should not show in the Camera media source UI (only the device name is shown instead)?

cgtobi commented:
I'll give it a try later. Do you want this to merge in the current state though and open a issue/PR for the customization topic or solve it in this PR?

codyc1515 commented:
I think merge it. I will raise another issue for how these are treated more broadly. It seems like we should get an overall consensus on how media source entities are displayed for all integrations (a direction). The original issue being that the name did not display at all (a bug) is now resolved.

joostlek commented:
Uhm, this is not a change we want to make. We want to move every integration to `has_entity_name = True`. Why can't it work with this?

joostlek commented:
If I understand correctly we are now implementing a workaround for something that should be fixed elsewhere. This should be reverted.

codyc1515 commented:
I’m not following. Do you have a suggestion for how to fix this? I followed the documentation which requires us to use this flag (not the deprecated way).
